Add Kubernetes operator webook server as DNS target

nitisht opened the pull request:
## Description
This PR adds a DNS target that ensures to update an entry
into Kubernetes operator when a bucket is created or deleted.

## Motivation and Context
See https://github.com/minio/operator/pull/264 & https://github.com/minio/operator/pull/273 for details.

## How to test this PR?

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

kerneltime commented on cmd/web-handlers.go:
```diff
@@ -198,6 +198,21 @@ func (web *webAPIHandlers) MakeBucket(r *http.Request, args *MakeBucketArgs, rep
 
+	if globalOperatorDNSConfig != nil {
+		// Proceed to creating a bucket.
+		if err := objectAPI.MakeBucketWithLocation(ctx, args.BucketName, opts); err != nil {
```
The operator is idempotent if the same bucket is created twice it returns 200, it might be better to create the service and then create the bucket. This way, there is no clean up needed.

kerneltime commented on cmd/bucket-handlers.go:
```diff
@@ -618,6 +618,15 @@ func (api objectAPIHandlers) PutBucketHandler(w http.ResponseWriter, r *http.Req
 		return
 	}
 
+	// Make a DNS entry in Kubernetes Operator
```
Same as below, the bucket creation on the operator is idempotent, so it is easier to create service then create the bucket and there is no need to clean up.

harshavardhana reviewed (changes requested):
@nitisht I would prefer a common model here i.e use globalDNSConfig this is one of the reasons why there is dns.Store implementation as an interface which you can follow such that we can return dns.Store which doesn't need to change any other code elsewhere.

harshavardhana commented on cmd/config/etcd/dns/operator_dns.go:
```diff
@@ -0,0 +1,167 @@
+/*
+ * MinIO Cloud Storage, (C) 2018,2020 MinIO, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
```

harshavardhana commented on cmd/config/etcd/dns/operator_dns.go:
```diff
@@ -0,0 +1,167 @@
+)
+
+func (c *OperatorDNS) endpointSuffix(bucket string, delete bool) string {
+	if delete {
```
use `url.Values` to construct this not manually adding `?`

harshavardhana commented on cmd/config/operator/operator.go:
```diff
@@ -0,0 +1,138 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
```

harshavardhana commented on pkg/env/web_env.go:
```diff
@@ -163,7 +163,7 @@ func Environ() []string {
 func LookupEnv(key string) (string, bool) {
 	v, ok := os.LookupEnv(key)
-	if ok && strings.HasPrefix(v, webEnvScheme) {
+	if ok && key == "MINIO_ARGS" {
```
Lookup ENV is generic is this for testing?

nitisht replied:
Lookup ENV is not generic, it assumes a GET is needed on every env variable that starts with `env` which is not the case

harshavardhana replied:
it is the case it is a generic implementation @nitisht the webhook server needs to implement the relevant part i.e to provide the expected ENV 

This change introduces bugs, this is not how MinIO side of the code should behave.

harshavardhana commented on cmd/config/operator/operator.go:
```diff
@@ -0,0 +1,138 @@
+		return "", "", "", err
+	}
+
+	switch u.Scheme {
```
You don't need to do any of this @nitisht all you are supposed to do is 

```
env.Get(EnvOperatorEndpoint, kvs.Get(endpoint))
```

This will automatically fetch the right endpoint by talking to remote operator if the `env` is `env:` you shouldn't be parsing `env` endpoint yourself.

nitisht replied:
This doesn't work because, if the prefix is `env` env.Get  calls `getEnvValueFromHTTP` which attempts a `GET` request on the endpoint. While that is fine for `MINIO_ARGS`. `GET` is not supported for `MINIO_OPERATOR_ENDPOINT`.

harshavardhana replied:
you should add that in operator @nitisht the endpoint is meant to get ENVs - you cannot monkey patch it on the client.

nitisht replied:
Why is a GET needed on bucketservice on operator?

nitisht replied:
what does that achieve?

harshavardhana replied:
Since you don't know the value of the env you should get it from the remote endpoint.

harshavardhana commented on pkg/env/web_env.go:
```diff
@@ -36,8 +36,8 @@ import (
+	WebEnvSchemeSecure = "env+tls"
 )
 
 var (
```
We don't need to expose these constants outside.

nitisht committed:
Add Kubernetes operator webook server as DNS target

This PR adds a DNS target that ensures to update an entry
into Kubernetes operator when a bucket is created or deleted.

See minio/operator#264 for details.

kerneltime commented on cmd/config/etcd/dns/operator_dns.go:
```diff
@@ -0,0 +1,200 @@
+			return err
+		}
+	}
+	if resp.StatusCode != http.StatusOK {
```
Should we also accept `StatusCreated`

harshavardhana committed:
Merge branch 'master' into bucket-service
